tests/FileSystemTests.cpp

```diff
@@ -773,6 +776,66 @@ TEST_CASE("FileSystemTests::CreateFileW helpers", "[filesystem]")
 }
 
 #endif
+TEST_CASE("FileSystemTest::FolderChangeReader destructor does not hang", "[filesystem]")
+{
+    wil::unique_cotaskmem_string testRootTmp;
+    REQUIRE_SUCCEEDED(wil::ExpandEnvironmentStringsW(L"%TEMP%\\wil_test_filesystem", testRootTmp));
+    std::wstring testRootDir = testRootTmp.get();
+    std::wstring testFile = testRootDir + L"\\test.dat";
+    bool deleteDir = false;
+    wil::unique_event_nothrow opCompletedEv; 
+    wil::unique_event_nothrow wilReaderNotify;
+
+    REQUIRE_SUCCEEDED(opCompletedEv.create(wil::EventOptions::None));
+    REQUIRE_SUCCEEDED(wilReaderNotify.create(wil::EventOptions::ManualReset));
+
+    REQUIRE_FALSE(DirectoryExists(testRootDir.c_str()));
+    REQUIRE_SUCCEEDED(wil::CreateDirectoryDeepNoThrow(testRootDir.c_str()));
+    REQUIRE(DirectoryExists(testRootDir.c_str()));
+
+    /**
+     * Move the operation to a new thread. 
+     * The destructor of unique_folder_change_reader might hang. If this happens, 
+     * we want to report an test error instead of hanging forever. 
+     * Initialize the reader in current thread to make sure there is no race condition with test 
+     * creating files
+     */
+    auto reader = wil::make_folder_change_reader_nothrow(testRootDir.c_str(), false, wil::FolderChangeEvents::All,
+    [&](wil::FolderChangeEvent, PCWSTR)
+    {
+        if (deleteDir)
+            RemoveDirectoryW(testRootDir.c_str());
+
+        SetEvent(opCompletedEv.get());
+    });
+    auto readerThread = std::thread([&wilReaderNotify, r = std::move(reader)]() mutable {
+
+        /**
+         * moving here to make sure the reader is destructed on a different thread 
+         */
+        auto r2 = std::move(r);
+        WaitForSingleObject(wilReaderNotify.get(), INFINITE);
+    });
 
+    wil::unique_hfile testFileOut(::CreateFileW(testFile.c_str(), GENERIC_ALL,
+        FILE_SHARE_READ, nullptr, CREATE_ALWAYS, 0, nullptr));
+    REQUIRE(testFileOut);
+    testFileOut.reset();
+    WaitForSingleObject(opCompletedEv.get(), INFINITE);
+
+    deleteDir = true;
+    REQUIRE(DeleteFileW(testFile.c_str()));
+    WaitForSingleObject(opCompletedEv.get(), INFINITE);
+    std::this_thread::sleep_for(std::chrono::seconds(5)); //enough time for the StartIO call to fail

```

Comment:
I am not really happy with this sleep here, but without it, there might be a race between calling the destructor and calling again `StartIo`

Reply:
How reliable is the repro if you: (1) decrease the timeout (e.g. maybe to something like 100ms), or (2) instead do something like `std::this_thread::yield`. If the issue reproduces (without this fix) a decent percentage of time (let's say >50% but maybe even less than that), IMO that's still decent enough that we'll catch regressions pretty quickly

Reply:
Wild thought: is the `m_tpIo` accessible off the `unique_folder_change_reader_nothrow` type? If so, I'm curious if the thread can call `WaitForThreadpoolIoCallbacks` itself. That would ensure that the callback - in its entirety - has completed, which includes the call to `StartIo`.

Reply:
Looks like the `get` function in the `unique_storage` class is public, so this seems possible

Reply:
> Wild thought: is the `m_tpIo` accessible off the `unique_folder_change_reader_nothrow` type? If so, I'm curious if the thread can call `WaitForThreadpoolIoCallbacks` itself. That would ensure that the callback - in its entirety - has completed, which includes the call to `StartIo`.

Curious if you gave this a try. I also wouldn't be opposed to adding a macro that conditionalizes making members public vs private for unit tests

Reply:
Does not seem to be possible with the current implementation. Calling
>PTP_IO readerRaw = reader.get()->m_tpIo;

gives me

> wil\include\wil\resource.h(347,107): error C2338: static_assert failed: 'get(): the raw handle value is not available for this resource class

Reply:
Ahh, I see now that `folder_change_reader_state_resource_policy` is marked as `pointer_access_none`. In theory we could do something like `#if WIL_TESTING` to make `pointer_access_none` act like `pointer_access_all`, _but_ that also sounds like something that we can address in the future